src/Parser.js

```diff
@@ -638,21 +641,25 @@ Parser.prototype.parseSpecialGroup = function(pos, mode, outerMode, optional) {
         this.expect(closeBrace, optional ? "]" : "}");
         return new ParseFuncOrArgument(
             new ParseResult(
-                new ParseNode(mode, data, outerMode),
+                new ParseNode(innerMode, data, outerMode),
                 closeBrace.position),
             false);
-    } else if (mode === "text") {
+    } else if (innerMode === "text") {
         // text mode is special because it should ignore the whitespace before
         // it
         var whitespace = this.lexer.lex(pos, "whitespace");
         pos = whitespace.position;
     }
 
+    this.mode = innerMode;
```

Comment:
The reason for doing the assignment down here is that `innerMode` can only be `text` or `math` by the time we get here.  Is that correct?  This could probably use a comment.